### PR TITLE
Configurable buildifier and buildozer executables

### PR DESCRIPTION
**Pull request by ignasl**
The development base was the bazelbuild/intellij repository, so some of the changes done in wix-playground/intellij are not properly reflected and integrated. This could be another PR.

Buildozer and buildifier paths are made configurable using Settings dialog box. Defaults are set to simply `buildozer` and `buildifier` so that system would pick up the one that is found anywhere in the PATH.

Fallback scheme for `buildozer`:
- Attempt to invoke the configured one
- If configured one fails and it is not the same as default (not found, etc.) invoke the default one in path (`buildozer`)
- If default one fails, use string concatenation to generate build file

At the moment there is no fallback scheme for buildifier yet. Fallback scheme for buildifier would require design change.

### CODE REVIEW

File: base/src/com/google/idea/blaze/base/lang/buildfile/actions/BuildFileModifierImpl.java [hunk 1]
```diff
@@ -270,9 +277,46 @@ private PsiElement createCommaElement(Project project) {
...
 
-    private PsiElement createRule(Project project, Kind ruleKind, String ruleName) {
-    String text =
-        Joiner.on("\n").join(ruleKind.toString() + "(", "    name = \"" + ruleName + "\"", ")");
+  private static String run(String[] command) throws IOException, InterruptedException {
+    return run(command, "");
+  }
+
+  private static String run(String[] command, String input) throws IOException, InterruptedException {
+    ProcessBuilder builder = new ProcessBuilder(command);
```
Comment (hmemcpy) [inline base/src/com/google/idea/blaze/base/lang/buildfile/actions/BuildFileModifierImpl.java]: Take a look at [OSProcessHandler](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-api/src/com/intellij/execution/process/OSProcessHandler.java) and [OutputListener](https://github.com/JetBrains/intellij-community/blob/master/platform/lang-impl/src/com/intellij/execution/OutputListener.java) - two IntelliJ utils that give you a wrapper over process execution.

It's used in various places, I'm also using it for the Wix Bazel plugin, see: https://github.com/wix-private/wix-intellij-plugin/blob/b912dcf13faf19b72c22451c14dcfd45e80b8b25/src/main/scala/com/wixpress/execution/package.scala#L59

File: base/src/com/google/idea/blaze/base/settings/BlazeUserSettings.java [hunk 2]
```diff
@@ -249,6 +254,38 @@ public void setBazelBinaryPath(String bazelBinaryPath) {
...
+  public void setUseBuiltInBuildifier(boolean useBuiltInBuildifier) {
+    this.useBuiltInBuildifier = useBuiltInBuildifier;
+  }
+
+  public String getBuildifierBinaryPath() {
+    return StringUtil.defaultIfEmpty(buildifierBinaryPath, DEFAULT_BUILDIFIER_PATH).trim();
+  }
+
+  public void setBuildifierBinaryPath(String buildifierBinaryPath) {
+    this.buildifierBinaryPath = StringUtil.defaultIfEmpty(buildifierBinaryPath, DEFAULT_BUILDIFIER_PATH).trim();;
```
Comment (hmemcpy) [inline base/src/com/google/idea/blaze/base/settings/BlazeUserSettings.java]: Extra semicolon

File: base/src/com/google/idea/blaze/base/settings/BlazeUserSettings.java [hunk 3]
```diff
@@ -249,6 +254,38 @@ public void setBazelBinaryPath(String bazelBinaryPath) {
...
+  }
+
+  public String getBuildozerBinaryPath() {
+    return StringUtil.defaultIfEmpty(buildozerBinaryPath, DEFAULT_BUILDOZER_PATH).trim();
+  }
+
+  public void setBuildozerBinaryPath(String buildozerBinaryPath) {
+    this.buildozerBinaryPath = StringUtil.defaultIfEmpty(buildozerBinaryPath, DEFAULT_BUILDOZER_PATH).trim();;
```
Comment (hmemcpy) [inline base/src/com/google/idea/blaze/base/settings/BlazeUserSettings.java]: Here too

File: base/src/com/google/idea/blaze/base/lang/buildfile/actions/BuildFileModifierImpl.java (hunk 1 above)
Comment (ignasl) [reply base/src/com/google/idea/blaze/base/lang/buildfile/actions/BuildFileModifierImpl.java]: Replaced ProcessBuilder with OSProcessHander

File: base/src/com/google/idea/blaze/base/settings/BlazeUserSettings.java (hunk 2 above)
Comment (ignasl) [reply base/src/com/google/idea/blaze/base/settings/BlazeUserSettings.java]: Removed

File: base/src/com/google/idea/blaze/base/settings/BlazeUserSettings.java (hunk 3 above)
Comment (ignasl) [reply base/src/com/google/idea/blaze/base/settings/BlazeUserSettings.java]: Removed

**Comment by hmemcpy** [comment]
I'm importing these changes into our plugin. I've asked @ignasl to close this PR, as well as the one upstream.

**Pull request closed without merging**